Enable structured binding return for Experimental::partition_space()

tcclevenger opened the pull request:
Addresses https://github.com/kokkos/kokkos/issues/8100. In case of weights being provided as compile-time variable number of inputs, return a `std::array<ExecSpace, sizeof...(Args)>` instead of a vector. This allows structured binding of return array
```
auto [exec0, exec1] = Experimental::partition_space(ExecSpace, 1, 1);
```

tcclevenger committed:
Enable structured binding return for Experimental::partition_space()

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented on core/src/Cuda/Kokkos_Cuda_Instance.cpp:
```diff
@@ -758,14 +758,12 @@ std::map<int, std::mutex> CudaInternal::constantMemMutexPerDevice     = {};
-            &stream)));
-    instances[s] = Cuda(stream, ManageStream::yes);
-  }
+void Kokkos::Impl::create_Cuda_instance(Cuda &instance) {
+  cudaStream_t stream;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(
+      (instance.impl_internal_space_instance()->cuda_stream_create_wrapper(
+          &stream)));
+  instance = Cuda(stream, ManageStream::yes);
```
Similar in HIP, since this is just an impl function, I didn't want to worry about having multiple that could handle array and vector, so I just had it handle a single instance and in CUDA's `partition_space()` we call this function in a loop over the new exec spaces.

tcclevenger commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -135,15 +135,15 @@ namespace Impl {
 // relative weight
 template <typename T>
-inline std::vector<OpenMP> create_OpenMP_instances(
+inline std::vector<int> calculate_omp_pool_sizes(
```
Changed this to just calculate pool sizes, instead of returning `OpenMP` instances since `partition_space` expects an array and a vector version of the instances.

maartenarnst commented:
This question may be a bit out of scope of this Pr, but just wondering if this function has the intended behavior with respect to the device id?

tcclevenger commented:
> This question may be a bit out of scope of this Pr, but just wondering if this function has the intended behavior with respect to the device id?

It will take the device ID of the default device. It may be a good idea to add some backend specific tests for the execution spaces returned by `partition_space()` (device ids for Cuda HIP, thread pool size for OpenMP, etc)

crtrott commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -131,14 +131,14 @@ TEST(TEST_CATEGORY, partitioning_by_args) {
   ASSERT_EQ(int(instances.size()), 2);
-  test_partitioning(instances);
+  test_partitioning(instances[0], instances[1]);
 }
```
Add a test which uses structured binding. Also lets add a
```
static_assert(std::is_same_v<decltype(instances), std::array<TEST_EXECSPACE, 2>>);
```
here and the corresponding one in the vector version.

crtrott reviewed (dismissed):
Looks pretty good just some test addition (i.e. test the structured binding and check the return type).

maartenarnst commented:
> > This question may be a bit out of scope of this Pr, but just wondering if this function has the intended behavior with respect to the device id?
> 
> It will take the device ID of the default device. It may be a good idea to add some backend specific tests for the execution spaces returned by `partition_space()` (device ids for Cuda HIP, thread pool size for OpenMP, etc)

Indeed, that’s also how I’m reading the current behavior. But I’m just wondering whether that’s really the behavior we expect when we "partition" a space? I.e., when we "partition" a space  associated with a given device, is it expected behavior that the returned spaces may actually be associated with a different device?

masterleinad commented:
> It will take the device ID of the default device. It may be a good idea to add some backend specific tests for the execution spaces returned by partition_space() (device ids for Cuda HIP, thread pool size for OpenMP, etc)

For `SYCL` we are using the same `context` and `device` for the new instances as the one passed in already.

masterleinad commented:
```
/home/runner/work/kokkos/kokkos/kokkos/core/unit_test/TestGraph.hpp:649:18: error: captured structured bindings are a C++20 extension [-Werror,-Wc++20-extensions]
  649 |         policy_t(exec_0, 0, 1),
      |                  ^
/home/runner/work/kokkos/kokkos/kokkos/core/unit_test/TestGraph.hpp:632:15: note: 'exec_0' declared here
  632 |   const auto [exec_0, exec_1, exec_2, exec_3] =
      |               ^
[ 24%] Building CXX object core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/serial/TestSerial_ViewCtorDimMatch.cpp.o
/home/runner/work/kokkos/kokkos/kokkos/core/unit_test/TestGraph.hpp:656:18: error: captured structured bindings are a C++20 extension [-Werror,-Wc++20-extensions]
  656 |         policy_t(exec_1, 0, 1),
      |                  ^
/home/runner/work/kokkos/kokkos/kokkos/core/unit_test/TestGraph.hpp:632:23: note: 'exec_1' declared here
  632 |   const auto [exec_0, exec_1, exec_2, exec_3] =
      |                       ^
```

dalg24 commented on core/src/Serial/Kokkos_Serial.hpp:
```diff
@@ -273,13 +273,12 @@ struct MemorySpaceAccess<Kokkos::Serial::memory_space,
-  std::vector<Serial> instances;
-  instances.reserve(sizeof...(Args));
-  std::generate_n(std::back_inserter(instances), sizeof...(Args),
+  std::array<Serial, sizeof...(Args)> instances;
```
This code is not equivalent.  We construct N default exec spaces and re-assign which is less efficient (not sure it is critical though).

Try
```suggestion
template <class... Args>
std::array<Serial, sizeof...(Args)> partition_space(const Serial&, Weights... ignored) {
  static_assert((... && std::is_arithmetic_v<Args>));
  return {(ignored, Serial(NewInstance{}), ...};
```

(same comment everywhere)

dalg24 reviewed (commented):
I think that it would be better to promote all weights to a `std::array` in a central place and have backend implementations take it as input, to guarantee that we always handle them the same and don't duplicate code.

tcclevenger commented:
Thanks for the feedback! Will be back to work Wednesday 5/28 to answer comments.

tcclevenger committed:
Remove structured binding return from TestGraph.hpp

Many of these instances are captured inside a lambda, which is not a
feature until C++20

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Avoid creating default constructed exec spaces in partition_space

For version returning std::array, use folded expression
((args, return_exec_space()) , ...)
to return an array inline.

For version returning std::vector, use vector::reserve and
vector::emplace_back.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Update TestExecSpacePartitioning.hpp

- Check that exec spaces have same device as the one used to construct
the partition spaces
- Static assert the array return type
- Add test for structured binding return type

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Add static asserts for weights in HPX backend

This bring the impl in line with all other backends

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
OpenMP: Check that thread pool size of default space is sum of partition spaces

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -163,24 +163,47 @@ inline std::vector<OpenMP> create_OpenMP_instances(
+  const auto pool_sizes =
+      Impl::calculate_omp_pool_sizes(main_instance, weights);
+
+  // Create array of OpenMP instances from pool sizes
+  auto create_partitions = [&pool_sizes]<std::size_t... Indices>(
+                               std::index_sequence<Indices...>) {
+    return std::array<OpenMP, sizeof...(Args)>{OpenMP(pool_sizes[Indices])...};
+  };
+
+  return create_partitions(std::make_index_sequence<sizeof...(Args)>{});
```
@dalg24 This is the only backend where it wasn't obvious to me that I could use folded expressions. What I've done might not be the most efficiently written, let me know if you have improvements/dislike the approach.

tcclevenger committed:
Move create_{Cuda|HIP}_instance() right above partition_space

This is the only place it's used, getting around a build error when
attempting to return the space in the original function.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Instance.hpp:
```diff
@@ -373,26 +371,40 @@ namespace Experimental {
-  std::vector<Cuda> instances(sizeof...(Args));
-  Kokkos::Impl::create_Cuda_instances(instances);
-  return instances;
+  return {((ignored, Impl::create_cuda_space(cuda_space)), ...)};
```
I would prefer readability over perfromance considerations or brevity here and use create an `std::array` with default instances first and the assign new ones (like you had initially). After all the creation of a new execution space instances should be much more expensive then creating a copy of the default execution space instance.

tcclevenger replied:
This new impl is causing build warnings for unused elements for the `ignored` variable, and I don't think it will look any nicer to fix those.So I think I agree and will revert.

dalg24 replied:
I would have suppressed the warning.  I disagree but won't block.

tcclevenger committed:
Revert back to allowing std::array to construct default spaces

The fold expression code looked too complex and lead to build warnings.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

masterleinad commented on core/src/SYCL/Kokkos_SYCL.hpp:
```diff
@@ -145,18 +145,19 @@ struct DeviceTypeTraits<Kokkos::SYCL> {
+
+  std::array<SYCL, sizeof...(Args)> instances;
+  for (auto& in : instances) {
+    in = sycl::queue(context, device, sycl::property::queue::in_order());
```
```suggestion
    in = SYCL(sycl::queue(context, device, sycl::property::queue::in_order()));
```

tcclevenger committed:
Fix SYCL errors

Was not constructing an execution space previously, just returning constructor input arg.

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Signed-off-by: tcclevenger <tccleve@sandia.gov>

dalg24 commented on core/src/Cuda/Kokkos_Cuda_Instance.hpp:
```diff
@@ -373,26 +371,43 @@ namespace Experimental {
+  KOKKOS_IMPL_CUDA_SAFE_CALL(
+      (base_space.impl_internal_space_instance()->cuda_stream_create_wrapper(
+          &stream)));
+  return Cuda(stream, Kokkos::Impl::ManageStream::yes);
```
```suggestion
  return Cuda(stream, ManageStream::yes);
```

tcclevenger replied:
This is in `Kokkos::Experimental::Impl` now, not `Kokkos::Impl` so if won't find `ManageStream`. I can move it back, but my personal preference is `Experimental::Impl::create_cuda_space()` since this is the only place where we use it currently and it matches the helper function for omp pool size calc.

maartenarnst commented:
> > > This question may be a bit out of scope of this Pr, but just wondering if this function has the intended behavior with respect to the device id?
> > 
> > 
> > It will take the device ID of the default device. It may be a good idea to add some backend specific tests for the execution spaces returned by `partition_space()` (device ids for Cuda HIP, thread pool size for OpenMP, etc)
> 
> Indeed, that’s also how I’m reading the current behavior. But I’m just wondering whether that’s really the behavior we expect when we "partition" a space? I.e., when we "partition" a space associated with a given device, is it expected behavior that the returned spaces may actually be associated with a different device?

@tcclevenger, would you mind creating an issue so that we can keep track of the issue of how this "partitioning" should behave wrt to the device? The behavior currently being different between cuda/hip and sycl, as @masterleinad had brought up, is probably something to mention in the issue as well. Thanks!

tcclevenger commented:
@maartenarnst It should be the same now for Cuda/HIP/SYCL, where the Cuda/HIP streams are created on the same device as the input space (not the default device as was before).

maartenarnst commented:
> @maartenarnst It should be the same now for Cuda/HIP/SYCL, where the Cuda/HIP streams are created on the same device as the input space (not the default device as was before).

@tcclevenger OK! In that case, we probably have something interesting to add to the PR description :)

tcclevenger commented:
> > @maartenarnst It should be the same now for Cuda/HIP/SYCL, where the Cuda/HIP streams are created on the same device as the input space (not the default device as was before).
> 
> @tcclevenger OK! In that case, we probably have something interesting to add to the PR description :)

Can do. Also I will update the comments today for each backend as well.

tcclevenger committed:
Centralize partition_space() for all exec spaces

Now, each backend needs only define impl_partition_space() given an
array (or vector) of instances, vector of weights, and a base instance.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented:
I attempted to reduce some duplicate code by consolidating all `partition_space()` into `Kokkos_Core.hpp`, and each backend is responsible for defining a custom `Experimental::Impl::imp_partition_space()`.

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -50,16 +50,22 @@ void check_distinctive([[maybe_unused]] ExecSpace exec1,
   if constexpr (std::is_same_v<ExecSpace, Kokkos::OpenMP>) {
     ASSERT_NE(exec1, exec2);
+    ASSERT_EQ(ExecSpace().impl_thread_pool_size(),
+              exec1.impl_thread_pool_size() + exec2.impl_thread_pool_size());
```
Why are we not using `ExecutionSpace::concurrency()`?

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -50,16 +50,22 @@ void check_distinctive([[maybe_unused]] ExecSpace exec1,
   if constexpr (std::is_same_v<ExecSpace, Kokkos::Cuda>) {
     ASSERT_NE(exec1.cuda_stream(), exec2.cuda_stream());
+    ASSERT_EQ(exec1.cuda_device(), ExecSpace().cuda_device());
+    ASSERT_EQ(exec2.cuda_device(), ExecSpace().cuda_device());
```
Assert that neither exec1 nor exec2 stream is the same than the stream from the origin argument exec

dalg24 replied:
Now I realize that you don;t have access to it in that function :/
You may ignore

dalg24 commented on core/src/SYCL/Kokkos_SYCL.hpp:
```diff
@@ -143,44 +143,19 @@ struct DeviceTypeTraits<Kokkos::SYCL> {
+  sycl::context context = base_instance.sycl_queue().get_context();
+  sycl::device device   = base_instance.sycl_queue().get_device();
+  for (auto& in : instances) {
+    in = SYCL(sycl::queue(context, device, sycl::property::queue::in_order()));
```
Probably not the place to handle it but isn't that in-order property something that should depend on whether KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES is defined?

masterleinad replied:
Yeah, I was going to do that in a follow-up.

dalg24 commented on core/src/Kokkos_Core.hpp:
```diff
@@ -288,25 +288,41 @@ class KOKKOS_ATTRIBUTE_NODISCARD ScopeGuard {
-  std::vector<ExecSpace> instances(sizeof...(Args));
-  for (int s = 0; s < int(sizeof...(Args)); s++) instances[s] = space;
+
+  std::vector<std::common_type_t<Args...>> weights = {args...};
```
I like the new approach where we handle weights in a centralized place.
I am wondering if we would want to go further and decide on 1 internal representation (that is choose one type) so that backends can have a non-templated implementation.

tcclevenger replied:
I elected to have the backends return vector, and `std::array partition_space()` converts vector -> array before returning.

dalg24 commented on core/src/Kokkos_Core.hpp:
```diff
@@ -316,7 +332,7 @@ std::vector<ExecSpace> partition_space(ExecSpace const& space,
       "Kokkos Error: partitioning arguments must be integers or floats");
 
   std::vector<ExecSpace> instances(weights.size());
```
I would have created an empty vector and reserved storage.
https://github.com/kokkos/kokkos/pull/8114#discussion_r2121174573
I am not blocking but you will need approval from someone else.

tcclevenger replied:
Each backend now uses reserve.

tcclevenger committed:
Use OpenMP::concurrency instead of impl_ function

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Have backends always return std::vector<Space> for impl_partition_space

In Kokkos_Core.hpp, convert to array for that specialization.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Avoid default constructing array when converting from vector

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented on core/src/Kokkos_Core.hpp:
```diff
@@ -288,25 +288,58 @@ class KOKKOS_ATTRIBUTE_NODISCARD ScopeGuard {
+  return instances;
+}
+
+// Helper function for converting std::vector -> std::array that avoids
+// declaring an std::array which is default constructed.
+template <class ExecSpace, size_t... Indices>
+std::array<ExecSpace, sizeof...(Indices)> create_partitions_array(
+    const std::vector<ExecSpace>& instances_vec,
+    std::index_sequence<Indices...>) {
+  return std::array<ExecSpace, sizeof...(Indices)>{instances_vec[Indices]...};
+}
```
@masterleinad @dalg24 I switched back to avoiding default constructing array before copying vector elements. A few differences from previous version
1. Now each backend isn't responsible for doing this, but it's centralized here, so anyone trying to say "I wonder what cuda does to partition..." isn't going to be met with complex code that may be tough to decipher.
2. This method _shouldn't_ create compiler errors like the fold expressions were. I also think it's slightly easier to read.
3. I added comments, so even if someone doesn't find this clear, they will know what is happening.

Let me know thoughts or disagreements.

masterleinad replied:
Either is fine with me.

dalg24 replied:
Note that now instead of incrementing the reference counter of the default instance we do it for the newly created child instances.  L311 is making copies of the vector elements.

tcclevenger replied:
Yeah, I didn't think about that. Then in that case I'd prefer the other way since it's more readable:
```
std::array<ExecSpace, sizeof...(Args)> instances;
std::copy(instances_vec.begin(), instances_vec.end(), instances.begin());
```
Unless there is another way I am not aware of to get around this?

dalg24 replied:
As I understand it, the default-initialization of the std::array will still yield the exec space default constructor to be called, meaning we'd be back to incrementing the reference counter of the default instance.

nliber replied:
Yes.  Everything in the std::array gets default-initialized, so that calls the default constructor on each instance of ExecSpace.

tcclevenger committed:
Use impl_internal_space_instance()->thread_pool_size() instead of concurrency()

Concurrency depends on omp_level, whereas partition_space only
guarantees internal thread pool size.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -51,7 +51,8 @@ void check_distinctive([[maybe_unused]] ExecSpace exec1,
     ASSERT_EQ(ExecSpace().concurrency(),
-              exec1.concurrency() + exec2.concurrency());
+              exec1.impl_internal_space_instance()->thread_pool_size() +
+                  exec2.impl_internal_space_instance()->thread_pool_size());
```
@Rombur can you please look into this and explain why we cannot use concurrency

Rombur replied:
When we wrote space partitioning for OpenMP the idea was that the new execution space would be used inside a parallel region. This was inspired by the old `partition_master`. When you are inside a parallel region, the concurrency is the same as the `thread_pool_size` . Outside of the parallel region, the concurrency is the number of threads available. I looked back at the PR,  I can't tell if this was a design decision or not.

dalg24 replied:
OK let's add a big fat FIXME_OPENMP comment and investigate this elsewhere

tcclevenger replied:
Added.

tcclevenger committed:
Copy instaces_vec to std::array using std::copy

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Add FIXME_OPENMP explaining why we can't use concurrency

Also, forgot to use thread_pool_size() for default instance, so change that here.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented:
@dalg24 The HPX fail is not in a test which uses partition space, so I think it is unrelated. I may relaunch the test to see if it is deterministic.

masterleinad commented:
> @dalg24 The HPX fail is not in a test which uses partition space, so I think it is unrelated. I may relaunch the test to see if it is deterministic.

The same failure also appears in https://github.com/kokkos/kokkos/actions/runs/15637408196/job/44056047925?pr=8173.

tcclevenger commented:
Added "General Enhancement" entry in changelog https://github.com/kokkos/kokkos/issues/7866
